### PR TITLE
fix(components): [tour] buttonProps & isSameProps errors

### LINKED ISSUE #15673: [Component] [tour-step] Tour组件bug
Bug Type: **`Component`**

## Environment

- Vue Version: `3.4.15`
- Element Plus Version: `2.5.3`
- Browser / OS: `windows`
- Build Tool: `Vite`

### Related Component

- `el-tour-step`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGVsLWJ1dHRvbiB0eXBlPVwicHJpbWFyeVwiIEBjbGljaz1cIm9wZW4gPSB0cnVlXCI+QmVnaW4gVG91cjwvZWwtYnV0dG9uPlxuXG4gIDxlbC1kaXZpZGVyIC8+XG5cbiAgPGVsLXNwYWNlPlxuICAgIDxlbC1idXR0b24gcmVmPVwicmVmMVwiPnVwbG9hZDwvZWwtYnV0dG9uPlxuICAgIDxlbC1idXR0b24gcmVmPVwicmVmMlwiIHR5cGU9XCJwcmltYXJ5XCI+c2F2ZTwvZWwtYnV0dG9uPlxuICAgIDxlbC1idXR0b24gcmVmPVwicmVmM1wiIDppY29uPVwiTW9yZUZpbGxlZFwiIC8+XG4gIDwvZWwtc3BhY2U+XG5cbiAgPGVsLXRvdXIgdi1tb2RlbD1cIm9wZW5cIj5cbiAgICA8ZWwtdG91ci1zdGVwXG4gICAgICA6dGFyZ2V0PVwicmVmMT8uJGVsXCJcbiAgICAgIHRpdGxlPVwiVXBsb2FkIEZpbGVcIlxuICAgICAgZGVzY3JpcHRpb249XCJQdXQgeW91IGZpbGVzIGhlcmUuXCJcbiAgICAvPlxuICAgIDxlbC10b3VyLXN0ZXBcbiAgICAgIDp0YXJnZXQ9XCJyZWYyPy4kZWxcIlxuICAgICAgdGl0bGU9XCJTYXZlXCJcbiAgICAgIGRlc2NyaXB0aW9uPVwiU2F2ZSB5b3VyIGNoYW5nZXNcIlxuICAgIC8+XG4gICAgPGVsLXRvdXItc3RlcFxuICAgICAgOnRhcmdldD1cInJlZjM/LiRlbFwiXG4gICAgICB0aXRsZT1cIk90aGVyIEFjdGlvbnNcIlxuICAgICAgZGVzY3JpcHRpb249XCJDbGljayB0byBzZWUgb3RoZXJcIlxuICAgIC8+XG4gICAgPHRlbXBsYXRlICNpbmRpY2F0b3JzPVwieyBjdXJyZW50LCB0b3RhbCB9XCI+XG4gICAgICA8c3Bhbj57eyBjdXJyZW50ICsgMSB9fSAvIHt7IHRvdGFsIH19PC9zcGFuPlxuICAgIDwvdGVtcGxhdGU+XG4gIDwvZWwtdG91cj5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgc2V0dXAgbGFuZz1cInRzXCI+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5pbXBvcnQgeyBNb3JlRmlsbGVkIH0gZnJvbSAnQGVsZW1lbnQtcGx1cy9pY29ucy12dWUnXG5pbXBvcnQgdHlwZSB7IEJ1dHRvbkluc3RhbmNlIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5jb25zdCByZWYxID0gcmVmPEJ1dHRvbkluc3RhbmNlPigpXG5jb25zdCByZWYyID0gcmVmPEJ1dHRvbkluc3RhbmNlPigpXG5jb25zdCByZWYzID0gcmVmPEJ1dHRvbkluc3RhbmNlPigpXG5cbmNvbnN0IG9wZW4gPSByZWYoZmFsc2UpXG48L3NjcmlwdD5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcImFsbG93SnNcIjogdHJ1ZSxcbiAgICBcImNoZWNrSnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsIl9vIjp7fX0=)

### Steps to reproduce

给非最后一个el-tour-step添加nextButtonProps属性，children值复核规则即可

### What is Expected?

下一步按钮展示为children内容

### What is actually happening?

展示的内容为finish

### Additional comments

tour-step组件的下一步按钮有一个判断，代码为
nextButtonProps?.children ?? current === total - 1
            ? t('el.tour.finish')
            : t('el.tour.next')
它会导致当nextButtonProps?.children为true时始终展示t('el.tour.finish')，因为空值合并运算符优先级比三元运算符优先级更高，因此先执行了 nextButtonProps?.children ?? current === total - 1 运算，导致传入children时三元运算符运算结果始终为true，应当修改为
nextButtonProps?.children ?? (current === total - 1
            ? t('el.tour.finish')
            : t('el.tour.next'))

### PR DESCRIPTION
**Pull request by Fuphoenixes**
Please make sure these boxes are checked before submitting your PR, thank you!
fixed https://github.com/element-plus/element-plus/issues/15673
fixed https://github.com/element-plus/element-plus/pull/15633
- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by Fuphoenixes** [comment]
@kooriookami cc

**Comment by Fuphoenixes** [comment]
image
I found that in jsx mode, there is still a "Maximum recursive updates exceeded" error, which means that the way of judging whether the sub-component props have changed is not possible. I reconstructed the way of writing here, instead of comparing the sub-component props, directly through provide & inject passes the props of the child component to the parent component to solve this problem.
我发现，在jsx模式下，还是会有 “Maximum recursive updates exceeded” 错误，说明判断子组件props是否变化这种写法不行，我重构了一下这里的写法，不在比对子组件props,直接通过依赖注入将子组件的props传给父组件解决了这个问题。

**Comment by kooriookami** [comment]
I searched `Maximum recursive updates exceeded` in Vue's issues. They said it caused by `side effects`.

**Comment by Fuphoenixes** [comment]
vue3.3版本的`Maximum recursive updates exceeded`错误还只是是 `warn`，但是vue3.4后重构了这一块，之后的这个错误类型变成了`Error`,会导致页面直接死掉，所以最近报这个问题的多了。

**Comment by Fuphoenixes** [comment]
这里对比子组件props时，由于“nextButtonProps”是对象类型的，所以就一致被判断为不相同，然后就进入了重复渲染。
其实之前这么写就不会抛错，template写法下可能有静态提升，所以没抛错。
image

**Comment by kooriookami** [comment]
Many components throw this error after Vue 3.4. It's too much trouble.

File: packages/components/tour/src/step.ts [hunk 1]
```diff
@@ -73,6 +74,7 @@ export const tourStepProps = buildProps({
    */
   scrollIntoViewOptions: {
     type: definePropType<boolean | ScrollIntoViewOptions>([Boolean, Object]),
+    default: undefined,
```
Comment (kooriookami) [inline packages/components/tour/src/step.ts:77]: Why do we need to add undefined?

File: packages/components/tour/src/step.vue [hunk 2]
```diff
@@ -98,12 +100,29 @@ const {
   onFinish: tourOnFinish,
   onChange,
 } = inject(tourKey)!
+props.showArrow
```
Comment (kooriookami) [inline packages/components/tour/src/step.vue:103]: What's this single expression?

File: packages/components/tour/src/step.vue [hunk 3]
```diff
@@ -98,12 +100,29 @@ const {
...
 
 const mergedShowClose = computed(() => props.showClose ?? showClose.value)
 const mergedCloseIcon = computed(
   () => props.closeIcon ?? closeIcon.value ?? Close
 )
 
+const filterButtonProps = (btnProps?: TourBtnProps) => {
+  if (!btnProps) return
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { children, onClick, ...others } = btnProps
```
Comment (kooriookami) [inline packages/components/tour/src/step.vue:122]: Why no-unused-vars?

Comment (Fuphoenixes) [reply packages/components/tour/src/step.vue]: I'll change it

File: packages/components/tour/src/step.ts (hunk 1 above)
Comment (Fuphoenixes) [reply packages/components/tour/src/step.ts:77]: Boolean 类型的 props，不传就是 false，这样就会覆盖掉父组件的默认配置了

Comment (Fuphoenixes) [reply packages/components/tour/src/step.ts:77]: image

Comment (kooriookami) [reply packages/components/tour/src/step.ts:77]: If you move Object in front of Boolean, is that useful?

Comment (Fuphoenixes) [reply packages/components/tour/src/step.ts:77]: unuseful

Comment (kooriookami) [reply packages/components/tour/src/step.ts:77]: okay

**Review by kooriookami — APPROVED** [review]

**Comment by kooriookami** [comment]
Is it ready to merge?

**Comment by Fuphoenixes** [comment]
yeah

**Pull request merged**